storage: Treat "precandidates" the same as "candidates" in replica GC

a-robinson opened the pull request:
We have long had code that relied on eagerly adding to the GC queue
replicas that are in the raft candidate state because it's often a
sign that the replica is no longer a part of the range. Now that
pre-vote is enabled, we need to do the same for the precandidate
state to ensure that replicas are GCed in a timely fashion.

Release note (bug fix): garbage collect old replicas in a more timely
fashion after a node has been offline for a long time (this bug only
exists in recent v2.0 alpha/beta releases, not in v1.1).

Fixes #23994, assuming this works in practice on the cluster involved there. @jordanlewis it looks like you're running a custom build there, otherwise I'd try this myself on the node in question. Can you try applying this patch to your branch and restarting jordan-tpcc-new:13 with it?

It'd be nice to also add a higher-level test for this situation of a node coming back online after being declared dead and the replicas for a range having been moved around the cluster. It's not going to be a simple test, though.

a-robinson committed:
storage: Treat "precandidates" the same as "candidates" in replica GC

We have long had code that relied on eagerly adding to the GC queue
replicas that are in the raft candidate state because it's often a
sign that the replica is no longer a part of the range. Now that
pre-vote is enabled, we need to do the same for the precandidate
state to ensure that replicas are GCed in a timely fashion.

Release note (bug fix): garbage collect old replicas in a more timely
fashion after a node has been offline for a long time (this bug only
exists in recent v2.0 alpha/beta releases, not in v1.1).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23995)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This needs a test, or perhaps file an issue that this needs to be tested. One option would be to pass the RaftState to `replicaGCShouldQueueImpl`. 

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23995#-:-L7pkVYHtbCNQqvjTPum:bpxu6db)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
> One option would be to pass the RaftState to replicaGCShouldQueueImpl.

Yeah, although if the code/test was written that way today, it still wouldn't have protected us from this regression. The problem isn't that we stopped detecting the `StateCandidate` raft status, it's that a new raft status was introduced that means the same thing in this context. Only a higher-level test would have caught the regression in behavior.

a-robinson commented:
Well, this is quite slow, but it appears to work. It's slow because the replicas don't enter the precandidate state until they're used, which doesn't happen until the replicate queue tries to grab their lease. And the replicate queue is taking 1 minute per replica due to the context deadline, so it's going to take `(2750 replicas) * (1 minute/replica) = 45.8 hours` to get through the entire backlog.

It'd be nice if `Replica`'s write code path better handled the fact that all of its peers return non-retryable errors, so that we could short circuit all the waiting.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but there's another missing piece to the puzzle. If the error were ReplicaTooOldError instead of RaftGroupDeletedError, Store.HandleRaftResponse would add the replica to the GC queue immediately, which should short-circuit the 1m timeout in the replication queue. 

We'd get ReplicaTooOldError if we talked to any current member of the range. In this case we're so far out of date (and the cluster is large enough) that none of our direct peers are still members of the range, so they can't tell us for certain that we're too old. We can't treat the two errors as equivalent because RaftGroupDeletedError doesn't convey as much certainty, but I think we could still add a case for RaftGroupDeletedError in HandleRaftResponse that adds the replica to the replica GC queue. 

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23995#-:-L7q6oM-5wM4kIFB87n8:b-lrmcz5)*
<!-- Sent from Reviewable.io -->

a-robinson committed:
storage: Add to replica GC queue in response to RaftGroupDeletedErrors

We believe this will help a node clear out its old replicas faster when
it comes online after being offline for so long that many of the
cluster's ranges have had their replicas shuffled around onto different
nodes.

Release note: None

a-robinson commented:
Yeah, the meaning of `RaftGroupDeletedError` is pretty different from `ReplicaTooOldError`, but I suppose that adding the replica to the replica GC queue erroneously isn't terribly expensive.

I'll test out a repro of the issue locally, since Jordan's cluster has been gone for over a day now.

bdarnell commented:
New commit looks good to me. The cost of incorrectly adding to the GC queue is basically one consistent read to the meta ranges, which could theoretically become a  hotspot if this became common but shouldn't really matter.

a-robinson commented:
The new commit very quickly cleared out a lot of the replicas on the node that was down, **but** it only cleared out the replicas for which the node has previously been the leaseholder for (or for which it didn't think it knew who the leaseholder was).

For all the other ranges that had previously had an epoch-based lease on another node that is still healthy, the downed node assumes that those leases are still valid since the other nodes' epochs haven't increased. Thus, all of those ranges remain in their dormant state.

This isn't the end of the world, and it's better than the failure mode from #23994 because there aren't any pending queue actions or crazy queue processing times, and I assume that restarting the other nodes to increment their epochs will fix it, but it's a little annoying.

a-robinson commented:
Just as an update:
1. Quickly restarting one of the nodes did not cause more of the replicas to get GC'ed.
2. This investigation has been temporarily paused due to https://github.com/cockroachdb/cockroach/issues/24023, which I want to figure out while I have the cluster in this state.

a-robinson commented:
So restarting the nodes does cause the remaining replicas to get GC'ed. It didn't work in my comment above due to #24023 affecting the cluster.

@bdarnell are you ok with this behavior? Basically, after being dead for a long time, the node will clean up all replicas for which the leaseholder's epoch has changed, but not those for which the leaseholder's epoch hasn't changed (until the epoch eventually gets incremented). This can leave a bunch of dormant, unusable replicas on the restarted node for theoretically forever if the other nodes were to never restart.

My take is that this is definitely good enough for v2.0, and pretty low priority for 2.1.

bdarnell commented:
> My take is that this is definitely good enough for v2.0, and pretty low priority to change for 2.1.

Agreed.

a-robinson commented:
Opened https://github.com/cockroachdb/cockroach/issues/24067 and https://github.com/cockroachdb/cockroach/issues/24068 to track the follow-up work.
